### ISSUE TITLE
Output formats fails to retain JATS `<title>` and `<label>` when nested inside `<statement>`

### ISSUE THREAD

**Issue by coryschires** [jgm/pandoc#8718]
## Problem
When converting from JATS to Markdown (or HTML and probably other outputs), both `<label>` and `<title>` tags are ignored when nested within `<statement>`.

Given the following (abridged) JATS XML:

```xml
<statement>
  <label>Hypothesis 1</label>
  <p>We hypothesize the following</p>
</statement>
```

The following command:

```bash
pandoc statement-with-label.xml -f jats -t markdown -o statement-with-label.md
```

Will produce to following Markdown:

```markdown
#

We hypothesize the following
```

## Solution / Expected Behavior
Instead, I would expect the label (or title) to be retained:

```markdown
# Hypothesis 1

We hypothesize the following
```

### Pandoc Version
```
pandoc 3.1
Features: +server +lua
Scripting engine: Lua 5.4
```

#### statement-with-label.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE article PUBLIC "-//NLM//DTD Journal Publishing DTD v3.0 20080202//EN" "journalpublishing3.dtd">
<article article-type="research-article" dtd-version="3.0" xml:lang="en" xmlns:mml="http://www.w3.org/1998/Math/MathML" xmlns:xlink="http://www.w3.org/1999/xlink">
  <body>
    <sec>
      <statement>
        <label>Hypothesis 1</label>
        <p>We hypothesize the following</p>
      </statement>
    </sec>
  </body>
</article>
```

#### statement-with-title.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE article PUBLIC "-//NLM//DTD Journal Publishing DTD v3.0 20080202//EN" "journalpublishing3.dtd">
<article article-type="research-article" dtd-version="3.0" xml:lang="en" xmlns:mml="http://www.w3.org/1998/Math/MathML" xmlns:xlink="http://www.w3.org/1999/xlink">
  <body>
    <sec>
      <statement>
        <title>Hypothesis 1</title>
        <p>We hypothesize the following</p>
      </statement>
    </sec>
  </body>
</article>
```

### Commands to reproduce
```bash
# Markdown as output
pandoc statement-with-label.xml -f jats -t markdown -o statement-with-label.md
pandoc statement-with-title.xml -f jats -t markdown -o statement-with-title.md

# HTML as output
pandoc statement-with-label.xml -f jats -t html -o statement-with-label.html
pandoc statement-with-title.xml -f jats -t html -o statement-with-title.html
```

---

Thanks for the help! If the fix is easy, we might be able to contribute a PR. Just point us in the right direction.

**Comment by kamoe** [comment]
The way the parseBlock function is written in the JATS reader, it would seem to me that this will be the case for all elements containing <label> and <title>, not only <statement> since lines 185-186 filter them out:

`        "title" -> return mempty -- processed by header
        "label" -> return mempty -- processed by header`

This will be for all elements, except for the <section> element, which deals with <label> and <title> specifically, in lines 317-334:

`sect n = do isbook <- gets jatsBook
                     let n' = if isbook || n == 0 then n + 1 else n
                     labelText <- case filterChild (named "label") e of
                                    Just t -> (<> ("." <> space)) <$>
                                              getInlines t
                                    Nothing -> return mempty
                     headerText <- case filterChild (named "title") e `mplus`
                                        (filterChild (named "info") e >>=
                                            filterChild (named "title")) of
                                      Just t  -> (labelText <>) <$>
                                                  getInlines t
                                      Nothing -> return mempty
                     oldN <- gets jatsSectionLevel
                     modify $ \st -> st{ jatsSectionLevel = n }
                     b <- getBlocks e
                     let ident = attrValue "id" e
                     modify $ \st -> st{ jatsSectionLevel = oldN }
                     return $ headerWith (ident,[],[]) n' headerText <> b`

(invoked from line 180):
`"sec" -> gets jatsSectionLevel >>= sect . (+1)`

So before defining an approach, I would like to ask, why was there the assumption that only <sec> needed these two elements?

A straightforward approach would be to expand on tarleb's suggestion, but for all elements affected, e.g. <bio>, <notes>, <glossary>, <kwd-group>, <app>, <app-group>, <back>, <abstract>, <ack>, which all contain <title> and <label> elements which get filtered out if not inside <sec>s.

**Comment by tarleb** [comment]
@hamishmack I know that this is a long shot for code that was written 6 years ago, but do you happen to remember the reasons for why things mentioned above are the way they are?

**Comment by hamishmack** [comment]
The code for JATS reader was based on the the DocBook one.  Perhaps this is just something we should have changed for JATS, but did not.

Here is the code from the DocBook.hs file at the time: https://github.com/jgm/pandoc/blob/5d3c9e56460165be452b672f12fc476e7a5ed3a9/src/Text/Pandoc/Readers/DocBook.hs#L893-L904

That code has changed only slightly and now look like this:
https://github.com/jgm/pandoc/blob/509cb9b8feae6798cb77bc35637297e9301d682e/src/Text/Pandoc/Readers/DocBook.hs#L1081-L1093

The two changes were https://github.com/jgm/pandoc/commit/12a35dd0d0f7363ad5b85ab859925113c65aa61f and https://github.com/jgm/pandoc/commit/40aa74badc2686b8b9a4ae7f836529cec2f4779b.

**Comment by coryschires** [comment]
+1 to https://github.com/jgm/pandoc/issues/8718#issuecomment-1489322480

I agree this is a more general problem than I initially understood. I encountered the same problem when working with [JATS's `<ack>` tag](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/ack.html). Like `<statement>`, `<ack>` can include nested `<label>` and / or `<title>` tags. These tags are similarly dropped when converting from JATS to MD. 

Also, agreeing with this https://github.com/jgm/pandoc/issues/8718#issuecomment-1489322480, I was able to work around this problem by converting the `<ack>` to `<sec>`. After making that change, the `<label>` / `<title>` are retained.

**Comment by kamoe** [comment]
@hamishmack I think the problem is not in the `sect` function but in `parseBlock`. 

If the JATs reader was based on the DocBook one, I can see how it could be assumed that section headers would handle all possible occurrences of `<title>` and `<label>`. But truth is, they don't. In the JATS reader, the two below lines completely write off the content of `<title>` and `<label>` elements outside of `<sec>` elements (which handle headers).

https://github.com/jgm/pandoc/blob/16f28ef5e945f3be14e05afb7d91f8adca18e49a/src/Text/Pandoc/Readers/JATS.hs#L185-L186

I can see that if we just remove these two lines, then we will duplicate `<title>` and `<label>` content in `<sec>` elements. Therefore, one improvement that is in order is to consider the specifics of the JATS specs that differ from the DocBook model, in particular, that there are 30+ different JATS elements different than `<sec>` that contain `<title>` and `<label>` elements, and therefore a big number of potential missing cases in the `parseBlock` function, which could all be modeled  after the cases of either `<sec>` or `<caption>`.

I am not sure if there is another more compact solution.

**Comment by jgm** [comment]
The question is how the `<title>` should be represented in these other elements.
In section, it becomes the contents of a Header element.

**Comment by coryschires** [comment]
:spiral_notepad: **Based on the [JATS Publishing Spec](https://jats.nlm.nih.gov/publishing/tag-library/1.3/) which differs slightly from the [JATS Archiving Spec](https://jats.nlm.nih.gov/archiving/)** :spiral_notepad:

## How should `<title>` be represented in these other elements?

**TLDR:** The JATS [`<title>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/title.html) tag should translate to a `Header` in all cases.

I've laid out all possible use cases in exhaustive detail. That said, I'll admit it's a little hard to give definitive answers in all cases because (a) the JATS spec is unfortunately loose and (b) many of these elements are not commonly seen in the wild (e.g. [`<question-preamble>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/question-preamble.html)), making it difficult to know how they are most commonly used. 

Where possible, I based my answers on examples from the spec or cases I found in the wild (e.g. on [PLOS ONE](https://journals.plos.org/plosone)). When I couldn't find examples, I simply used my best judgement. (I'm certainly not the world's premiere JATS expert but, fwiw, I am a member of the JATS Standing Committee so I'm not clueless).

The [`<title>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/title.html) tag may be contained in:

* [`<abstract>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/abstract.html)
    * Should render as `Header` element
* [`<ack>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/ack.html)
    * Should render as `Header` element
* [`<answer>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/answer.html)
    * Should render as `Header` element
* [`<answer-set>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/answer-set.html)
    * Should render as `Header` element
* [`<app>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/app.html)
    * Should render as `Header` element
* [`<app-group>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/app-group.html)
    * Should render as `Header` element
* [`<author-comment>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/author-comment.html)
    * Should render as `Header` element
* [`<author-notes>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/author-notes.html)
    * Should render as `Header` element
* [`<back>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/back.html)
    * Should render as `Header` element
* [`<bio>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/bio.html)
    * Should render as `Header` element
* [`<caption>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/caption.html)
    * Should render as `Header` element
* [`<def-list>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/def-list.html)
    * Should render as `Header` element
* [`<disp-quote>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/disp-quote.html)
    * Should render as `Header` element
* [`<explanation>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/explanation.html)
    * Should render as `Header` element
* [`<fn-group>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/fn-group.html)
    * Should render as `Header` element
* [`<glossary>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/glossary.html)
    * Should render as `Header` element
* [`<kwd-group>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/kwd-group.html)
    * Should render as `Header` element
* [`<list>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/list.html)
    * Should render as `Header` element
* [`<list-item>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/list-item.html)
    * Should render as `Header` element
* [`<notes>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/notes.html)
    * Should render as `Header` element
* [`<option>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/option.html)
    * Should render as `Header` element
* [`<question>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/question.html)
    * Should render as `Header` element
* [`<question-preamble>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/question-preamble.html)
    * Should render as `Header` element
* [`<question-wrap-group>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/question-wrap-group.html)
    * Should render as `Header` element
* [`<ref-list>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/ref-list.html)
    * Should render as `Header` element
* [`<sec>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/sec.html)
    * Should render as `Header` element
* [`<statement>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/statement.html)
    * Should render as `Header` element
* [`<supplement>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/supplement.html)
    * Should render as `Header` element
* [`<table-wrap-foot>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/table-wrap-foot.html)
    * Should render as `Header` element
* [`<trans-abstract>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/trans-abstract.html)
    * Should render as `Header` element
* [`<verse-group>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/verse-group.html)
    * Should render as `Header` element

## What's the difference between [`<label>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/label.html) and [`<title>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/title.html)?

**TLDR:** Both `<label>` and `<title>` should be converted into `Header`.

Same disclaimer as above: I'm doing my best to grapple with inherent (and unfortunate) ambiguity.

According to the JATS spec:

* [`<label>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/label.html) – Number and/or prefix word placed at the beginning of display elements (for example, equation, statement, figure).
* [`<title>`](https://jats.nlm.nih.gov/publishing/tag-library/1.3/element/title.html) – Heading or title for a structural element (for example, `<sec>`, `<app>`, `<boxed-text>`).

Furthermore, in most (perhaps all?) cases, an element (e.g. `<sec>`) can contain both a `<label>` and `<title>` – and they must appear in that order:

```xml
<sec>
  <label>3.</label>
  <title>Conclusions</title>
</sec>
```

However, it's also common to only have one of either `<label>` or `<title>`:

```xml
<sec>
  <title>Conclusions</title>
</sec>
```

```xml
<statement>
  <label>Hypothesis 1</label>
  <p>Buyer preferences for companies are influenced...</p>
</statement>
```

### So how to handle these `<label>` / `<title>` combinations?

**Given** a `<sec>` includes both a `<label>` and `<title>`
```xml
<sec>
  <label>3.</label>
  <title>Conclusions</title>
</sec>
```
**Then** collapse them into a single `Header`
```markdown
# 3. Conclusions
```

--- 

**Given** a `<statement>` includes only a `<label>`
```xml
<statement>
  <label>Hypothesis 1</label>
  <p>Buyer preferences for companies are influenced...</p>
</statement>
```
**Then** convert the `<label>` into a `Header`
```markdown
# Hypothesis 1

Buyer preferences for companies are influenced...
```

---
**Given** a `<sec>` includes only a `<title>`
```xml
<sec>
  <title>Conclusions</title>
</sec>
```
**Then** convert the `<title>` into a `Header`
```markdown
# Conclusions
```

Hope this helps clarify possible next steps!

**Comment by kamoe** [comment]
> So, which version should Pandoc target or prefer?
> Unfortunately, I don't think there's an obviously "right" answer to this question. 

Why not? Pandoc is not a validation tool. If it supports more elements than it "should" for any given document, it is not going to produce an error. It will parse the document correctly anyway.

As far as I can gather from the JATS reader source code, all Pandoc does is consider the cases of selected elements, builds Pandoc structures in the AST document for those elements; then continues parsing that elements's children recursively.  If an element does not have a case defined, then its children are parsed, and execution continues down the line, until all is left is plain text. So unless you specifically wipe an element off (as we are currently doing with `<title>` and `<label>`), no content will get ignored (It might not be pretty formatted, but it will be there in the final AST). That's it.

Therefore, supporting a specific JATS suit only means that enough significant elements proper to that suite have been included in the parsing cases, or in the helper functions. Now, as of the latest released version of Pandoc, the element `<X>`, which is only present in JATS Archival and Interchange, and not in either of the two other suites, has indeed a case defined in he `isBlock` function. That is why I gather that, currently, Pandoc supports, or at least tries to support, JATS Archiving and Interchange. 

This decision makes sense to me because, following the  logic of case coverage, if it supports the most complete one, it supports all the three JATS suites.

> As you point out, Archiving is the largest. But, fwiw, Publishing is the most commonly used (and I suspect by a wide margin).
> 
I am working towards [expanding the JATS reader to include support for BITS](https://github.com/jgm/pandoc/issues/7740), which is an extension of JATS Archival and Interchange. So even if the JATS reader did not support that now, when the BITS reader is ready, it will. And this does not mean it will not support the other two suites.

> 
> > I think this has clearly become an epic concerning <title> and  across all elements, not only a specific bug concerning the  element.
> 
> Agreed. We're happy to help with code changes. But at this point, I am unfortunately not sure about the next steps.

Same, I'm not sure if we can just propose changes, or at what point any given proposal gets the go ahead.

**Comment by jgm** [comment]
If you're both agreed on what should change, maybe you could give an executive summary here?  I'm not a user of JATS myself, so I'm a bit lost in the details!

**Comment by kamoe** [comment]
I believe the bottomline is, we need to write cases in the `parseBlock` function of the JATS reader, for the following elements:

`<abstract>`
`<ack>`
`<answer>`
`<answer-set>`
`<app>`
`<app-group>`
`<author-comment>`
`<author-notes>`
`<back>`
`<bio>`
`<def-list>`
`<disp-quote>`
`<explanation>`
`<fn-group>`
`<glossary>`
`<kwd-group>`
`<list>`
`<list-item>`
`<notes>`
`<option>`
`<question>`
`<question-preamble>`
`<question-wrap-group>`
`<ref-list>`
`<sec>`
`<statement>`
`<supplement>`
`<table-wrap-foot>`
`<trans-abstract>`
`<verse-group>`

A reasonable solution would be if we just treated all those elements the same way we treat `<sec>`, that is, create a subsection with a header that is one level higher than the current header:

https://github.com/jgm/pandoc/blob/16f28ef5e945f3be14e05afb7d91f8adca18e49a/src/Text/Pandoc/Readers/JATS.hs#L180

Because this solution addresses the issue for all elements that contain a `<title>`, we are not only addressing the original issue reported here, but also https://github.com/jgm/pandoc/issues/8364, and https://github.com/jgm/pandoc/issues/8365.

Do you agree, @coryschires?

**Comment by jgm** [comment]
Thanks. Currently we have
```hs
        "title" -> return mempty -- processed by header
```

Replacing this with something that parses the inline contents of the title element, emits a header at currentlevel + 1, and (unlike sect) does not modify currentlevel, would probably get you most of the way there. The only issue is that we'd have to continue processing "title" specially when it occurs in the initial info element.

**Comment by kamoe** [comment]
I just reviewed the mechanics of `label` and `title` in JATS when they co-occur, and it really seems like having them both on the same heading is imperative  (e.g. "Section 8 Technical background", where "Section 8" is the `label` and "Technical background" is the `title`).

**Comment by noahmalmed** [comment]
Thanks for all of the direction on this @kamoe ! I'm going to take a crack at it

**Comment by noahmalmed** [comment]
Got a fix up here: #8840 if anyone wants to take a look!

**Comment by kamoe** [comment]
@noahmalmed Great work, thank you! I have finished added all my comments, and have also asked a question on the PR wrt what the best way to proceed is to have this merged. Happy to help in any way I can.

**Comment by kamoe** [comment]
@noahmalmed I'm afraid there is a problem with the current solution tough. If you notice, it only solves the problem for `<title>`, but not for `<label>`. We are completely removing the `<label>` even if no `<title>` is present. I thought the rationale was to remove `<label>` if and only if also a `<title>` was present...

@coryschires What do you think?

**Comment by noahmalmed** [comment]
@kamoe Oh shoot, I guess I misunderstood. @jgm , was the idea to only suppress label when title was present or did we want to just fully suppress it?

**Comment by kamoe** [comment]
Just to clarify, I think it might or might not make sense to leave as is, I am just pointing out that we might need to retain a lone `<label>` in certain circumstances?

I'm thinking more in terms of current JATS practice, like the examples given above by @coryschires originally when they opened the issue. Would "Hypotesis 1" be something that be duplicated down the line somehow?
